Allow changing opt-in checkbox position on the checkout page  [MAILPOET-5807]

NeosinneR opened the pull request:
## Description

This PR adds a new setting to change the position of the opt-in checkbox on the checkout page. 

<img width="981" alt="Screenshot 2024-07-31 at 00 54 38" src="https://github.com/user-attachments/assets/33235e42-1ab6-445c-82f2-5256666a3305">

## Code review notes

I realized that it's not clear which checkout page we run tests against (block or shortcode), so I updated tests to explicitly use one or the other or both to increase test coverage. Especially with AutomateWoo, it turned out to be valuable because changing hooks caused the AutomateWoo opt-in checkbox to display, even if it shouldn't. 

## QA notes

1. Test that all opt-in checkbox positions are working as expected. 
2. Test that the opt-in checkbox on old installations remains positioned before terms and conditions.
3. Test that the opt-in checkbox on new installations is positioned after billing info. 
4. Test that the AutomateWoo checkbox is hidden when MailPoet's is displayed. 

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5807]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] 🚫 I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5807]: https://mailpoet.atlassian.net/browse/MAILPOET-5807?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

NeosinneR committed:
Use __() for translations in checkout-optin setting

[MAILPOET-5807]

NeosinneR committed:
Add opt-in position setting

[MAILPOET-5807]

NeosinneR committed:
Change the condition to check if opt-in on checkout is enabled to be easier to read

[MAILPOET-5807]

NeosinneR committed:
Use proper hook based on opt-in position setting

[MAILPOET-5807]

NeosinneR committed:
Allow setting opt-in message in tests

[MAILPOET-5807]

NeosinneR committed:
Differentiate between block and shortcode checkout in tests

[MAILPOET-5807]

NeosinneR committed:
Improve opt-in message test by checking both block and shortcode checkout pages

[MAILPOET-5807]

NeosinneR committed:
Improve AutomateWoo integartion tests by checking both block and shortcode checkout pages

[MAILPOET-5807]

NeosinneR committed:
Add acceptance tests for opt-in checkbox positions on shortcode checkout page

[MAILPOET-5807]

NeosinneR committed:
Fix hiding AutomateWoo opt-in checkbox when MailPoet opt-in is in different position

[MAILPOET-5807]

NeosinneR committed:
Test all combinations of opt-in checkbox positions on block and shortcode checkout with AutomateWoo

[MAILPOET-5807]

NeosinneR committed:
Add migration so existing active installations remain opt-in checkbox position

[MAILPOET-5807]

costasovo commented on mailpoet/lib/Migrations/App/Migration_20240730_212419_App.php:
```diff
@@ -0,0 +1,29 @@
+  public function run(): void {
+    $this->settings = $this->container->get(SettingsController::class);
+
+    // Skip if the installed version is newer than the release that preceded this migration, or if it's a fresh install
+    $currentlyInstalledVersion = (string)$this->settings->get('db_version', '4.58.1');
+    if (version_compare($currentlyInstalledVersion, '4.58.0', '>')) {
+      return;
+    }
+    // Skip if the opt-in checkbox is not enabled. When enabling, the user chooses
+    // the position of the opt-in checkbox on the same settings page, so no default
+    // value is necessary.
+    if (!$this->settings->get(Subscription::OPTIN_ENABLED_SETTING_NAME, false)) {
+      return;
+    }
+    $this->settings->set(Subscription::OPTIN_POSITION_SETTING_NAME, Hooks::OPTIN_POSITION_BEFORE_TERMS_AND_CONDITIONS);
+  }
```
Perhaps we could skip the version check and just check if the position setting is missing. 

```php
// value is not necessary or is set
if (
 !$this->settings->get(Subscription::OPTIN_ENABLED_SETTING_NAME, false)
 || $this->settings->get(Subscription::OPTIN_POSITION_SETTING_NAME, false)
 ) {
  return;
 }
 
 $this->settings->set(Subscription::OPTIN_POSITION_SETTING_NAME, Hooks::OPTIN_POSITION_BEFORE_TERMS_AND_CONDITIONS);
 ```
It is a tiny detail, but I like to avoid the version check conditions. They are always a bit risky as we may forget to update the version in case the PR merge is postponed. And we are already reading the settings.

NeosinneR replied:
Good call; I've updated the migration. It also made me realize, that I forgot to set the default value in Popuplator, so I also added that.

costasovo reviewed (commented):
@NeosinneR Great job!
Thank you for improving the tests! The checkout functionality is very important so it is great we cover all the combinations.

I have just a tiny suggestion.

NeosinneR committed:
Add default value for opt-in position to Populator

[MAILPOET-5807]

NeosinneR committed:
Update migration to only check for presence of the setting, without version check

[MAILPOET-5807]

costasovo reviewed (dismissed):
@NeosinneR Thank you for the updates!

NeosinneR committed:
Add tracking for opt-in position

[MAILPOET-5807]

Aschepikov commented:
@NeosinneR Hi everything works well but there is a tiny issue, reproducible with chrome and edge
![image](https://github.com/user-attachments/assets/4ad2a8ca-056e-48e9-adc5-f366998aa5f3)

NeosinneR committed:
Remove line-height from small variants of inputs

It was causing descender to be cut off. Removing it doesn't change the overall height of the input

[MAILPOET-5807]

NeosinneR commented:
@Aschepikov fixed in https://github.com/mailpoet/mailpoet/pull/5764/commits/a7cc7f34ed83db6fad7c83308cbedcbea7538470 :) I think it's small enough change not to ask for another code review, so once the build is green, feel free to approve and merge if it's okay by you

Aschepikov reviewed (approved):
GJ!
